crates/exchange-rate-oracle/src/lib.rs

```diff
@@ -259,30 +243,82 @@ pub mod pallet {
+            // this is only for the initial parachain state, before a single value has been fed
+            Self::report_oracle_offline(None);
+        }
+        let current_time = Self::get_current_time();
+
+        for (ref key, is_updated) in raw_values_updated {
+            if !is_updated && !Self::is_outdated(key, current_time) {
+                continue;
+            }
+
+            RawValuesUpdated::<T>::insert(key, false);
+
+            let mut raw_values: Vec<_> = RawValues::<T>::iter_prefix(key).map(|(_, value)| value).collect();
+            let min_timestamp = Self::get_current_time().saturating_sub(Self::get_max_delay());
+            raw_values.retain(|value| value.timestamp >= min_timestamp);
+            if raw_values.len() == 0 {
+                Self::report_oracle_offline(Some(key));
+            } else {
+                let valid_until = raw_values
+                    .iter()
+                    .map(|x| x.timestamp)
+                    .min()
+                    .map(|timestamp| timestamp + Self::get_max_delay())
+                    .unwrap_or_default(); // Unwrap will never fail, but if somehow it did, we retry next block
+
+                let mid_index = raw_values.len() / 2;
+                let (_, value, _) = raw_values.select_nth_unstable_by(mid_index as usize, |a, b| a.value.cmp(&b.value));
+
+                if Aggregate::<T>::get(key).is_none() {
+                    Self::recover_from_oracle_offline();
+                }
+
+                Aggregate::<T>::insert(key, value.value);
+                ValidUntil::<T>::insert(key, valid_until);
+            }
         }
     }
 
+    // public only for testing purposes
+    pub fn _feed_values(oracle: T::AccountId, values: Vec<(OracleKey, T::UnsignedFixedPoint)>) -> DispatchResult {
+        for (key, value) in values.iter() {
+            let timestamped = TimestampedValue {
+                timestamp: Self::get_current_time(),
+                value: value.clone(),
+            };
+            RawValues::<T>::insert(key, &oracle, timestamped);
+            RawValuesUpdated::<T>::insert(key, true);
+        }
+
+        Self::deposit_event(Event::<T>::FeedValues(oracle, values));
+
+        Ok(())
+    }
+
     /// Public getters
 
     /// Get the exchange rate in planck per satoshi
-    pub fn get_exchange_rate() -> Result<UnsignedFixedPoint<T>, DispatchError> {
-        let max_delay_passed = Self::is_max_delay_passed();
-        ensure!(!max_delay_passed, Error::<T>::MissingExchangeRate);
-        Ok(<ExchangeRate<T>>::get())
+    pub fn get_price(key: OracleKey) -> Result<UnsignedFixedPoint<T>, DispatchError> {
+        Aggregate::<T>::get(key).ok_or(Error::<T>::MissingExchangeRate.into())
     }
 
     pub fn wrapped_to_collateral(amount: Wrapped<T>) -> Result<Collateral<T>, DispatchError> {
-        let rate = Self::get_exchange_rate()?;
+        let rate = Self::get_price(OracleKey::ExchangeRate(CurrencyId::DOT))?;
```

Comment:
I'm okay with this if we eventually use the `currency_id` of the targeted vault, but it's a bit annoying to have this hard-coded here.

Reply:
Yea this is just temporary, when we move to multi collateral we will remove this